estimateAlpha

ChouaibB opened the pull request:
Initial draft implementation.

Commit:
draft

antagomir commented:
Related to #417

antagomir committed:
Merge branch 'master' into issue_417

antagomir commented on R/estimateAlphaWithRarefaction.R:
```diff
@@ -0,0 +1,84 @@
+                                         name = args.fun$index){
+    # checks
+    if(!.is_an_integer(nrounds)) {
+        stop("'nrounds' must be an interger.",
```
interger -> integer..

antagomir commented on R/estimateAlphaWithRarefaction.R:
```diff
@@ -0,0 +1,84 @@
+             call. = FALSE)
+    }
+    
+    # Generating seeds for every round
+    set.seed(seed)
+    SEEDS <- sample.int(10000, size = nrounds)
+    colData(x)[, name] <- lapply(seq(nrounds), function(i){
```
IMO it is entirely sufficient to set one seed before the entire lapply; setting step-specific seed is an overkill

TuomasBorman replied:
Regarding setting seed. We have this kind of solution in couple of functions, but this is not recommended way according to Bioconductor. Seed should be set outside the function (Both solutions are fine by me)

antagomir replied:
I also think it is best to set the seed outside of the function. Users still have all the control over this, and it simplifies the function.

antagomir commented on tests/testthat/test-10estimateAlphaWithRarefaction.R:
```diff
@@ -0,0 +1,12 @@
+test_that("Estimate Alpha Diversity Indices with Rarefaction", {
+    data(GlobalPatterns, package="mia")
+    tse <- GlobalPatterns
+    # Calculate the default Shannon index with 1 rarefaction round
+    tse <- estimateAlphaWithRarefaction(tse)
+    expect_true(any(grepl("shannon", colnames(colData(tse)))))
+
+    # Calculate the default observed richness with 10 rarefaction rounds
+    tse <- estimateAlphaWithRarefaction(tse, nrounds=10,
+     FUN=mia::estimateRichness, args.fun=list(index="observed"))
+    expect_true(any(grepl("observed", colnames(colData(tse)))))
+})
```
If possible, I would run the same tests than are run for estimateDiversity (and other diversity functions) but with and without the rarefaction argument.

antagomir commented:
I am tending to think that we should combine all these estimateXXX functions into a single estimateAlpha function as your PR suggest. The other estimateXXX functions could be turned into internal functions to simplify the interface. That estimateAlpha function can then have rarefaction as optional argument (including the user-specified rarefaction depth; by default the lowest read count among the samples).

Things to consider:
1) previously exported estimateXXX functions would need deprecation message when they are moved into internal functions
2) the diversity index names should have suffixes, like "shannon_diversity", "observed_richness" etc.
3) there is a mechanism available in R that can pick those argument names even when just the (unique) start of the index name is given, like "shannon" would be recognized to mean "shannon_diversity" so the users can use shorter names if they need to but the output need to use the full names (for instance Simpson index is available both for diversity and evenness, therefore it must be indicated which one it is).

If unclear, we can discuss more.

I have some further comments in mind after the next PR draft on this.

antagomir commented:
The function has to be changed into estimateAlpha, which then includes rarefaction as an optional argument

Commit:
Merge branch 'master' of https://github.com/microbiome/mia into issue_417

Commit:
draft

Commit:
Merge branch 'issue_417' of https://github.com/microbiome/mia into issue_417

antagomir commented:
An idea: the method name (like "shannon") could also say "diversity", then the function would return all "diversity" indices and if one asks for "dominance" then those etc (richness / evenness / diversity / dominance / rarity are the options and the indices are listed in those corresponding functions now). If no method is specific, then all indices would be returned by default.

TuomasBorman commented:
I think this is not the best way to implement this. Better way is to keep those estimateDiversity.R and other files and just add estimateAlpha.R with estimateAlpha functionality (remove documentation from necessary places)

The problem with moving all the code to one file is that it is a nightmare for someone that tries to check what has been changed. It looks like everything has

antagomir commented:
I agree. The old and extensively tested functions can be kept as is. They should be just converted into internal functions that are not exported, and then called from a wrapper called `estimateAlpha()`. This should be also faster to implement than rewriting the entire function.

ChouaibB commented:
Alright, I misunderstood the comment I will make the changes.

antagomir commented:
Always aim at minimal and straightfwd changes unless otherwise agreed.

If you could update this PR accordingly that would be great!

ChouaibB commented:
> An idea: the method name (like "shannon") could also say "diversity", then the function would return all "diversity" indices and if one asks for "dominance" then those etc (richness / evenness / diversity / dominance / rarity are the options and the indices are listed in those corresponding functions now). If no method is specific, then all indices would be returned by default.

I didn't understand, could you please explain more.

ChouaibB commented:
> > An idea: the method name (like "shannon") could also say "diversity", then the function would return all "diversity" indices and if one asks for "dominance" then those etc (richness / evenness / diversity / dominance / rarity are the options and the indices are listed in those corresponding functions now). If no method is specific, then all indices would be returned by default.
> 
> I didn't understand, could you please explain more.

Do you mean smt like this?
`
.get_indices <- function(measure) {
    switch(measure,
           "diversity" = c("coverage_diversity", "coverage",
                           "faith_diversity", "faith",
                           "fisher_diversity", "fisher",
                           "gini_simpson_diversity", "gini_simpson",
                           "inverse_simpson_diversity", "inverse_simpson",
                           "log_modulo_skewness_diversity", "log_modulo_skewness",
                           "shannon_diversity", "shannon"),
           "dominance" = c("absolute_dominance", "absolute",
                           "dbp_dominance", "dbp",
                           "core_abundance_dominance", "core_abundance",
                           "gini_dominance", "gini", 
                           "dmn_dominance", "dmn",
                           "relative_dominance", "relative",
                           "simpson_lambda_dominance", "simpson_lambda"),
           "evenness" = c("camargo_evenness", "camargo",
                          "pielou_evenness", "pielou",
                          "simpson_evenness",
                          "evar_evenness", "evar",
                          "bulla_evenness", "bulla"),
           "richness" = c("ace_richness", "ace",
                          "chao1_richness", "chao1",
                          "hill_richness", "hill",
                          "observed_richness", "observed"))
}
`

Commit:
restructured

Commit:
Merge branch 'master' of https://github.com/microbiome/mia into issue_417

Commit:
BPPARAM could be passed as ...

Commit:
supressed examples warnings + fixed ...

Commit:
supressed warning at example + vignette

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+                                             min_size=min(colSums(assay(x, "counts")),
+                                                          na.rm = TRUE),
+                                             verbose=FALSE),
+                               FUN=estimateDiversity,
```
should this rather be estimateAlpha?

ChouaibB replied:
not actually I should change it to .estimate_diversity

antagomir reviewed (commented):
Nice! Some comments added. Perhaps TB has more

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+#' @param assay.type the name of the assay used for
+#'   calculation of the sample-wise estimates.
+#'   
+#' @param assay_name a single \code{character} value for specifying which
+#'   assay to use for calculation.
+#'   (Please use \code{assay.type} instead. At some point \code{assay_name}
+#'   will be disabled.)
```
Do we even need that alias since we are removing it and this is a new function?

antagomir replied:
True. Alias is not needed for estimateAlpha. 

For other (earlier) diversity estimation functions maybe.

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+                          rarify=FALSE,
+                          seed = 123,
+                          nrounds=10,
+                          rarefaction_depth=min(colSums(assay(x, "counts")), na.rm = TRUE)){
```
rarefaction_depth could be NULL by default which then is calculated inside the function. --> that way user can run this with default settings by only specifying the assay.type

OR (maybe better if works) can you replace "counts" with assay.type? Does it work?

ChouaibB replied:
I will check that out

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+             call. = FALSE)
+    }
+    FUN <- NULL
+    if(index %in% .get_indices("diversity")) {
```
MIght be good to have a comment what this block does

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+    
+    if (rarify) {
+        .alpha_rarefaction(x, nrounds = nrounds, seed = seed,
+                           args.sub = list(assay.type=assay.type,
+                                           min_size=rarefaction_depth,
+                                           verbose=FALSE),
```
RUn BiocCheck::BiocCheck(), its seems that this indentation is not correct

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+
+.parse_name <- function(index, name, measure) {
+    # don't change name if defined by user
+    if (name==index) {
```
index can be a vector of indices? Does this then work?

ChouaibB replied:
Actually good point. I have made so far in such a way that it handles one index. I will upgrade it to handle multiple ones.

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+    return(x)
+}
+
+.parse_name <- function(index, name, measure) {
+    # don't change name if defined by user
+    if (name==index) {
+        if (measure %in% unlist(strsplit(index, "\\_"))) {
+            name = index
+        } else {
+            name = paste0(index, "_", measure)
+        }
+    } else {
+        return(name)
+    }
+}
+
```
Might be good to have comment what this does

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+    set.seed(seed)
+    colData(x)[, name] <- lapply(seq(nrounds), function(i){
+        x_sub <- do.call(subsampleCounts, args = c(list(x), args.sub))
+        suppressWarnings(x_sub <- do.call(FUN, args = c(list(x_sub),
```
Why warnings are suppressed? Add comment there. These should be avoided

ChouaibB replied:
Due to the deprication message added to the alpha functions, it needed to be supressed when called from estimateAlpha to pass the test.

TuomasBorman reviewed (changes requested):
COding style should more or less match between functions even though we have multiple developers. For example there is no many comments

ChouaibB commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+#' @export
+estimateAlpha <- function(x, assay.type = "counts", assay_name = NULL,
+                          index = c("coverage_diversity", "fisher_diversity",
+                                    "faith_diversity", "faith",
```
I thought that's what you requested, user can use both to refer to the alpha measure to be used?

ChouaibB replied:
it's removed

antagomir replied:
I think this is automatically done in R functions when the shorter name matches the full name.

ChouaibB commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+.parse_name <- function(index, name, measure) {
+    # don't change name if defined by user
+    if (name==index) {
+        if (measure %in% unlist(strsplit(index, "\\_"))) {
```
not working

antagomir replied:
ok

ChouaibB commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,248 @@
+                               index = c("camargo", "pielou", "simpson_evenness",
+                                         "evar", "bulla"),
+                               name = index, ...) {
+    estimateEvenness(x, assay.type = assay.type, assay_name = assay_name,
```
Yes, deprication message was added to those functions at their own R files.

ChouaibB commented on tests/testthat/test-10estimateAlpha.R:
```diff
@@ -0,0 +1,51 @@
+test_that("Estimate Alpha Diversity Indices with Rarefaction", {
```
ok.
I checked all indices of each alpha measure before implementing, and it turns out actually all are unique to their own functions.

ChouaibB commented on tests/testthat/test-10estimateAlpha.R:
```diff
@@ -0,0 +1,51 @@
+    expect_true(any(grepl("pielou_evenness", colnames(colData(tse)))))
+    # Calculate same index with 10 rarefaction rounds
+    tse <- estimateAlpha(tse, assay.type = "counts", index = "pielou",
+                         rarify=TRUE, nrounds=10, name="pielou_10")
```
ok, i will add one with 10 and another with 20.

antagomir replied:
and initalized with different random seeds

Commit:
fixes

Commit:
more fixes

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,247 @@
+#' rarify=TRUE, n.iter=10)
+#' 
+#' # Shows the estimated observed richness
+#' colData(tse)$observed_richness
```
same

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,247 @@
+                          name = index,
+                          ...,
+                          rarify=FALSE,
+                          seed = 123,
```
remove seed from the function

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,247 @@
+                          seed = 123,
+                          n.iter=10,
+                          rarefaction_depth=min(colSums(assay(x, assay.type)), na.rm = TRUE)){
+    # checks
```
more explanations needed; add a comment line for each substantial step below to help reader understand what it does. Check other functions for examples.

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,247 @@
+
+## Helper functions
+
+.get_indices <- function(measure) {
```
nice!

antagomir commented on R/estimateDiversity.R:
```diff
@@ -183,7 +183,9 @@
 #' "Faith",  "LogModSkewness")
```
does this function estimateDiversity now have a deprecation message?

ChouaibB replied:
yes

antagomir reviewed (commented):
JUst some rather minor comments.

antagomir commented:
After this PR is accepted, we need to remember update OMA.

ChouaibB commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,247 @@
+#' @param rarify logical scalar: Should the alpha diversity measures be estimated
+#'   using rarefaction? (default: \code{FALSE})
+#' 
+#' @param seed a single \code{integer} value as the seed used for the nround
+#'  rarefaction.
```
I thought the result of the discussion was that the seed is ok to keep outside? subsampling will use internally other seeds, and having a seed outside the call of subsampling will keep results reproducible?

antagomir replied:
Yes. Outside of the function entirely.

antagomir replied:
It will keep results reproducible if you set a seed, then call a function.

Commit:
fixes

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,245 @@
+            index[i] <- gsub("_richness", "", index[i])
+            FUN <- .estimate_richness
+        } else {
+            stop("'index' is coresponding to none of the alpha diversity indices.",
```
Instead of saying what it is not, say what it should be.

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,245 @@
+                                    ...,
+                                    name=name[i])
+        } else {
+            suppressWarnings(x <- do.call(FUN, args = c(list(x, assay.type=assay.type,
```
Add comment on what this is supposed to do

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,245 @@
+                                          assay.type="subsampled"),
+                               ...,
+                               name = args.fun$index) {
+    set.seed(seed)
```
seed still included despite the discussion otherwise?

ChouaibB replied:
Once again, the earlier discussion resulted in keeping the seed for reproducibility? the subsampling function does it internally for each iteration, so keeping a seed outside the function call will enable the user to reproduce results....?

antagomir replied:
Seed can be given outside the function and the results remain reproducible. 

Bioconductor recommends not having an additional seed in the functions.

I do not see why we need the seed inside function for reproducibility. You can test this yourself: apply the function two times without setting seed outside -> different results; set seed outside and immediately run the function -> same results every time

ChouaibB replied:
ok I checked the subsampling function, I see now that it would produce the same seed over and over. But then results are not reproducible for the whole process due to [link](https://github.com/microbiome/mia/blob/master/R/subsampleCounts.R#L87C28-L87C61) ... but guess that's the problem of the subsampling function. I remove it from here.

antagomir replied:
Yes, let us set seed outside of function if necessary, and remove from the function itself. Then the function result will remain reproducible.

This should always generate the same:

```
set.seed(24452)
estimateAlpha(...)
```

If this is not the case, prove me wrong.

ChouaibB replied:
I already said setting outside the function is reproducible. "Outside the function" was not clear at comments for me....
but setting one seed before the lapply is not going to make the rarefaction happen since same subsampling results will be produced in every iteration.....

ChouaibB replied:
That's why the communication was confusing after asking at the beginning to set one seed before the lapply...

antagomir replied:
I do not think that statment is true: if you set seed outside the function, and then run lapply within the function, the lapply will not produce the same subsampling results on every iteration. If it does, that is unexpected and I would like to get a reproducible example that demonstrates that.

ChouaibB replied:
Or perhaps I have cognitive problems too, which does not make me a good candidate for being a developer... IDK it's possible...

antagomir replied:
You can try it out: run lapply() in those alternative ways:

# One general seed before lapply
1) 
set.seed(123); res1a <- lapply(....)
set.seed(123); res1b <- lapply(....)
-> res1a and res1b will yield identical outcomes -> result is reproducible

# Same seed initialized again at each iteration
2) res2 <- lapply(...., function (x) {set.seed(123); ... })
-> this will yield different outcome than res1a or res1b because here the seed is reset to 123 at each new iteration instead of just continuing the seedings from where the previous iteration left

# Random seed per each iteration
3) res3 <- lapply(...., function (x) {set.seed(sample(1e6, 1)); ... })
-> this yields different outcome from res1a, res1b, res2 since it uses different (randomly generated) seed at every iteration

ChouaibB replied:
> I do not think that statment is true: if you set seed outside the function, and then run lapply within the function, the lapply will not produce the same subsampling results on every iteration. If it does, that is unexpected and I would like to get a reproducible example that demonstrates that.

Yes it won't, it will produced only the first iteration the same and the rest is not reproducible, and the question I had in mind is how would you make the whole iterations reproducible to have final reproducible results. That's why I still do think that generating a list of seeds and passing on each iteration the corresonding seed to subsampleCount is simple, safe and definitely reproducible...

ChouaibB replied:
ok, I see your point now, I didn't know that setting one seed before lapply will make reproducible result:
![image](https://github.com/microbiome/mia/assets/86964562/ac2d90eb-f09a-454c-9b42-edc0c383147c)

antagomir replied:
Yes, this is what I said: the result for the entire loop is reproducible when the seed it set before applying the loop.

antagomir replied:
Is this now clear?

ChouaibB replied:
Yes it is, I tested also outside the function and gives same results.
But what's honestly not clear is why you asked me to put it before the lapply at the begining then re-asked to remove it, and Tuomas also was ok with both solutions and you didn't comment? :) I have noticed this multiple times and it's the same story in every PR I make... I am not really motivated for this especially when I have other tasks at hand waiting for me to be done...  It's ok I understand I am not good enough of a developer, so perhaps I should do smt else instead? or go to coding 101. Maybe I could just open issues when encounter some and come here when I am good enough someday?

antagomir replied:
The discussion evolved. 

Initially I thought that before lapply is enough. Then Tuomas pointed out that Bioc does not recommend including seed in the function. It sounded reasonable, so I changed my thinking and concluded that we can remove the seed from the function entirely. I might have not communicated that clearly in a hurry as I mistakenly thought I stated it in the later comments that the seed should be placed outside of the function.

The other points we should discuss elsewhere than in the public Github issues, thanks.

antagomir replied:
My apologies, if I have not expressed the task with sufficient clarity, will try to improve on that now.

antagomir commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,245 @@
+        # check if suffix of the alpha indices if present at index
+        # otherwise keeping suffix as a name if name not defined by user.
+        if (measure %in% unlist(strsplit(index, "\\_"))) {
+            name = index
+            } else {
```
are the indentations ok here? 

also name=index should be name <- index here i think

Commit:
seed removed

Commit:
fix

Commit:
fix

TuomasBorman committed:
Merge branch 'master' into issue_417

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,245 @@
+                          n.iter=10,
+                          rarefaction.depth=max(colSums(assay(x, assay.type)), na.rm = TRUE)){
+    # checks
+    if(is.null(index) & any(!sapply(index, .is_non_empty_string))) {
```
You can use argument matching here. I think we have used in other functions. match.arg or something

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,245 @@
+                                    ...,
+                                    name=name[i])
+        } else {
+            # Estimate index without rarefaction
```
Why there is a warning? Add comment

TuomasBorman replied:
suppressing warnings should be avoided. Probably there is a good reason for that, but it might be good to add info why it is done

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,245 @@
+                          "observed_richness"))
+}
+
+.alpha_rarefaction <- function(x,
+                               n.iter=1L,
```
As a note, you could add line break after `function(` --> the code would look nicer

TuomasBorman commented on vignettes/mia.Rmd:
````diff
@@ -226,7 +226,7 @@ and results are stored automatically in `colData`. Selected indices can be
 
 ```{r}
-tse <- estimateDiversity(tse)
+suppressWarnings(tse <- estimateDiversity(tse))
````
You could use this new function

TuomasBorman reviewed (changes requested):
Looks good! I will check this in more detail (try the function) when I get my computer working

Commit:
Merge branch 'master' of https://github.com/microbiome/mia into issue_417

Commit:
up

Commit:
Merge branch 'issue_417' of https://github.com/microbiome/mia into issue_417

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,274 @@
+#' 
+#' @param x a \code{\link{SummarizedExperiment}} object.
+#' 
+#' @param assay.type the name of the assay used for
+#'   calculation of the sample-wise estimates.
+#'   
+#' @param index a \code{character} vector, specifying the alpha diversity indices
+#'   to be calculated
+#'   
+#' @param name a name for the column(s) of the colData the results should be
+#'   stored in. By default this will use the original names of the calculated
+#'   indices.
+#'   
+#' @param ... optional arguments.
+#' 
```
Add default values

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,274 @@
+              if(!.is_an_integer(n.iter)) {
+                  stop("'n.iter' must be an integer.", call. = FALSE)
+                  }
+              if(!(is.numeric(rarefaction.depth) & rarefaction.depth > 0)) {
```
is.numeric(rarefaction.depth) && rarefaction.depth > 0) --> use two &

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,274 @@
+                      name[i] <- .parse_name(index[i], name[i], "richness")
+                      index[i] <- gsub("_richness", "", index[i])
+                      FUN <- .estimate_richness
+                  } else {
+                  stop("'index' is coresponding to none of the alpha diversity indices.
+                  'index' should be one of: coverage_diversity, fisher_diversity,
+                  faith_diversity, gini_simpson_diversity,
+                  inverse_simpson_diversity, log_modulo_skewness_diversity,
+                  shannon_diversity, absolute_dominance, dbp_dominance,
+                  core_abundance_dominance, gini_dominance,
+                  dmn_dominance, relative_dominance, simpson_lambda_dominance,
+                  camargo_evenness, pielou_evenness, simpson_evenness,
+                  evar_evenness, bulla_evenness, ace_richness, chao1_richness,
+                  hill_richness or observed_richness.",
+                       call. = FALSE)
+                  }
```
Check already in input check that every index can be found. There you can also parse the name. --> fail fast

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,274 @@
+                                                        list(...))))
+        # Storing estimate results
+        colData(x_sub)[, args.fun$index, drop=FALSE]
+    }) %>% data.frame(.) %>% rowMeans(.) %>% data.frame(.)
```
You could split these to multiple steps. I mean first you could calculate the value to variable a, then do these modifications to it and then return

ChouaibB replied:
?
Is this not working?

TuomasBorman replied:
Yes, it is working. However, it is little bit hard to follow. I am thinking if we need to debug things, so it is easier that thinkgs are done in steps rsther than with one go

ChouaibB replied:
I do think it's easy to follow and it's straight forward. What is complicated there at the end of the day?

ChouaibB replied:
it is perfectly fine in my opinion, you could perhaps change i tin future fixes and commits or what so ever.. :)

ChouaibB replied:
I do think that using dplyr grammar for coding is already clear...

TuomasBorman replied:
As I said, this is good like this, but we have preferred base grammar, no big deal

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,274 @@
+    }
+}
+
+.estimate_diversity <- function(
+        x, assay.type = "counts",
+        index = c("coverage", "fisher", "gini_simpson",  "inverse_simpson",
+                  "log_modulo_skewness","shannon"),
+        name = index, ...) {
+    estimateDiversity(x, assay.type=assay.type, index=index, name=name, ...)
+}
+
+.estimate_dominance <- function(
+        x,
+        assay.type = "counts", 
+        index = c("absolute", "dbp", "core_abundance", "gini", "dmn",  "relative",
+                  "simpson_lambda"),
```
Move these functions under estimate*.R file. Move the functionality of estimateDiversity under .estimate_diversity. --> Then we can use these functions in the future internally without warnings

TuomasBorman replied:
Easiest and tidiest way is to just rename estimateDiversity with .estimate_diversity and create a new exported function estimateDiversity which utilizes this "old" function

Cannot see if you have already made changes because Im on my phone

TuomasBorman commented on R/estimateAlpha.R:
```diff
@@ -0,0 +1,274 @@
+#' rounds.
+#' 
+#' @param rarefaction.depth a \code{double} value as for the minimim size or 
+#' rarefaction.depth. (default: \code{min(colSums(assay(x, "counts")), na.rm = TRUE)})
+#' 
```
I think the default value should be NULL --> no rarefaction. It is more clear, I think

TuomasBorman replied:
Or mayne we could use TRUE/FALSE values and then numeric

TRUE --> default rarefsction with message that says the rarefaction depth

FALSE --> no rarefaction

Numeric --> use user-defined depth

TuomasBorman reviewed (changes requested):
Final comments.

Lastly, run BiocCheck::BiocCheck() and fix warnings/notes related to this function (Check the mia.BiocCheck folder and the file inside it)

ChouaibB commented:
> Final comments.
> 
> Lastly, run BiocCheck::BiocCheck() and fix warnings/notes related to this function (Check the mia.BiocCheck folder and the file inside it)

The earlier devtools::check() there were no error (ofc otherwise won't commit) and no warnings, few notes that regards other functions. But I could test with BiocCheck::BiocCheck() now.

TuomasBorman commented:
> > Final comments.
> > 
> > Lastly, run BiocCheck::BiocCheck() and fix warnings/notes related to this function (Check the mia.BiocCheck folder and the file inside it)
> 
> The earlier devtools::check() there were no error (ofc otherwise won't commit) and no warnings, few notes that regards other functions. But I could test with BiocCheck::BiocCheck() now.

Yes, bioccheck cheks the coding style also

Commit:
up

ChouaibB commented:
I couldn't get the BiocCheck installed but the results of the devtools::check are as follows (complete log attached):
![image](https://github.com/microbiome/mia/assets/86964562/bf7de6e2-2a78-460c-ab27-94c48c135240)
[00check.log](https://github.com/microbiome/mia/files/13226620/00check.log)
It seems to be there are no warning regarding the function but perhaps others.

Anyway, you could please do whatever you want with these commits and the issue.
Thanks and good luck! :)

TuomasBorman commented:
Ok, there was atleast some indentation problems (Bioc has a rule that indentation must be multiplication of 4 spaces)

But these are easy and fast to fix, thanks!

ChouaibB commented:
Thank you! good luck with all! :)

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'master' into issue_417

antagomir committed:
Merge branch 'master' into issue_417

antagomir commented:
up - if anything more needed from my side to close this PR let me know

TuomasBorman commented:
At least the name should be addAlpha to follow new naming scheme (related to that we can have a discussion about naming, I will send you msg)

antagomir committed:
Merge branch 'devel' into issue_417

antagomir commented:
@ChouaibB can you follow the suggestion from @TuomasBorman and update? Then actively follow-up on this until the PR is approved by us and merged successfully.

antagomir commented:
Related to #417

TuomasBorman committed:
Delete R/mia.BiocCheck/00BiocCheck.log

TuomasBorman commented:
This looks good, the name should be addAlpha- And also it might be best to move deprecated functions to deprecate.R file

TuomasBorman committed:
Merge branch 'devel' into issue_417

TuomasBorman committed:
Fix indentation

TuomasBorman committed:
Add comments

TuomasBorman committed:
sync

Merge branch 'issue_417' of github.com:microbiome/mia into issue_417

# Conflicts:
#	R/subsampleCounts.R
#	man/subsampleCounts.Rd

Commit:
rename estimateAlpha to addAlpha and deprecate estimate* functions

Commit:
merge devel

Merge branch 'devel' of https://github.com/microbiome/mia into issue_417

# Conflicts:
#	DESCRIPTION

antagomir commented:
Ok @ChouaibB let's take this up again and aim to finalize.

Commit:
fix + rm test at 10-test

Commit:
fixes related to test-8subsample

Commit:
fixed warnings

Commit:
@doctype --> @aliases

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+#' Estimate alpha diversity indices.
```
header line should not include period (Bioc guideline)

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+#' Estimate alpha diversity indices.
+#' 
+#' The function estimates alpha diversity indices optionally using of rarefaction,
```
Remove "The function" and capitalize estimates; 

"using of" -> "using";

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+#' Estimate alpha diversity indices.
+#' 
+#' The function estimates alpha diversity indices optionally using of rarefaction,
+#' then stores results at \code{\link{colData}}.
```
at -> in

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+#' 
+#' @param x a \code{\link{SummarizedExperiment}} object.
+#' 
+#' @param assay.type the name of the assay used for
+#'   calculation of the sample-wise estimates (default: \code{assay.type = "counts"}).
+#'   
+#' @param index a \code{character} vector, specifying the alpha diversity indices
+#'   to be calculated.
+#'   
+#' @param name a name for the column(s) of the colData the results should be
+#'   stored in. By default this will use the original names of the calculated
```
can we use @inheritParams for arguments that are shared with other functions?

ChouaibB replied:
umm.. I tried to check but couldn't spot a benefit here? since now all the estimate*alpha* functions are transferred to the current implementation? (kinda addAlpha now is the parent function?)

antagomir replied:
Yes, addAlpha is the main function.

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+#' # Shows the estimated observed richness
+#' tse$observed_richness
+#' 
+
```
remove this empty line

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+#' tse$shannon
+#'
+#' # Calculate observed richness with 10 rarefaction rounds
+#' tse <- addAlpha(tse, assay.type = "counts", index = "observed_richness",
+#' rarefaction.depth=min(colSums(assay(tse, "counts")), na.rm = TRUE), n.iter=10)
```
split on multiple lines, one line per argument for easier read

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+#' @name addAlpha
+#' @rdname addAlpha
+#' @export
+setGeneric(
+    "addAlpha", signature = c("x"), function(
+        x, assay.type = "counts", 
```
Check: is this PR suggesting an entirely new function "addAlpha", or is this modification of the already existing function "addAlpha" that did not have rarefaction before?

Ideally, this PR will extend the existing alpha diversity functionality and only add the parts that are needed to add rarefaction.

ChouaibB replied:
This is PR is about to add the `addAlpha` function and also depricate and change previous estimate*Alpha* functions.

antagomir replied:
Perfect.

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+            "coverage_diversity", "fisher_diversity", "faith_diversity",
+            "gini_simpson_diversity", "inverse_simpson_diversity",
+            "log_modulo_skewness_diversity", "shannon_diversity",
+            "absolute_dominance", "dbp_dominance", "core_abundance_dominance",
+            "gini_dominance", "dmn_dominance", "relative_dominance",
+            "simpson_lambda_dominance", "camargo_evenness", "pielou_evenness",
+            "simpson_evenness", "evar_evenness", "bulla_evenness",
```
fix intendations

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,229 @@
+            "coverage_diversity", "fisher_diversity", "faith_diversity",
+            "gini_simpson_diversity", "inverse_simpson_diversity",
+            "log_modulo_skewness_diversity", "shannon_diversity",
+            "absolute_dominance", "dbp_dominance", "core_abundance_dominance",
+            "gini_dominance", "dmn_dominance", "relative_dominance",
+            "simpson_lambda_dominance", "camargo_evenness", "pielou_evenness",
+            "simpson_evenness", "evar_evenness", "bulla_evenness",
```
fix intendations

antagomir commented on R/estimateEvenness.R:
```diff
@@ -114,28 +114,25 @@
-                              "bulla"),
-                    name = index, ...)
-               standardGeneric("estimateEvenness"))
+setGeneric(
+    ".estimate_evenness",signature = c("x"),
+    function(
+        x, assay.type = assay_name, assay_name = "counts",
+        index = c("pielou", "camargo", "simpson_evenness", "evar", "bulla"),
+        name = index, ...)
+        standardGeneric(".estimate_evenness"))
```
is there an actual change here? if not, then it would be better to keep the original format so we can focus the review on the actual changes. Same comment goes for all content.

ChouaibB replied:
yes: estimate_evenness --> .estimate_evenness

antagomir commented:
I will add some comments - wait until that before further changes.

TuomasBorman commented:
Also important to check that the code follows Biconductor guidelines.

1. BiocCheck::BiocCheck()
2. Fix issues related to these lines that are added by this PR (there are not many. I quickly looked that indentation was off in couple of lines)

Commit:
tmp commits

Commit:
resolve conflict+merge

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#' Estimate alpha diversity indices
+#' 
+#' The function estimates alpha diversity indices optionally using rarefaction,
```
Change: "The function estimates" -> "Estimate"

antagomir replied:
Remove the words: "optionally using rarefaction," (this will be explained later in arguments and @details field)

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#' 
+#' @param x a \code{\link{SummarizedExperiment}} object.
+#' 
+#' @param assay.type the name of the assay used for calculation of the
+#'   sample-wise estimates (default: \code{assay.type = "counts"}).
+#'   
+#' @param index a \code{character} vector, specifying the alpha diversity 
+#'   indices to be calculated.
+#'   
+#' @param name a name for the column(s) of the colData the results should be
+#'   stored in. By default this will use the original names of the calculated
```
We could use @inheritParams for : "assay.type" and "name" arguments from some other functions?

TuomasBorman replied:
We can do this to all manual pages at the same time

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#'   stored in. By default this will use the original names of the calculated
+#'   indices(By default: \code{name = index}).
+#'   
+#' @param ... optional arguments.
```
This "..." should be the last argument to explain (by convention).

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#' @param n.iter a single \code{integer} value for the number of rarefaction
+#'    rounds (By default: \code{n.iter = 10}).
+#' 
+#' @param rarefaction.depth a \code{double} value as for the minimim size or 
```
minimim -> minimum

antagomir replied:
If this is not given, does it use min(colSums(assay(tse, "counts")), na.rm = TRUE) by default?

ChouaibB replied:
It was using it before but now it's set to NULL by default.

antagomir replied:
Ok - what happens when rarefaction is activated, and this has its default value (NULL) - no rarefaction?

ChouaibB replied:
When NULL no rarefaction, when not NULL it is checked to be strictly positive number to activate rarefaction:
```
if( !is.null(rarefaction.depth) && 
            !(is.numeric(rarefaction.depth) && rarefaction.depth > 0)) {
            stop("'rarefaction.depth' must be a non-zero positive double.",
                call. = FALSE)
        }
```
I thought this was as requested before? I could try change it to the implementation suggested at the below comment if needed?

antagomir replied:
We have two rarefaction parameters: rarefaction.depth and n.iter; I think one of these should be enough to activate rarefaction, otherwise users might get confused if two different things need to be activated to make it work. Perhaps an intuitive solution would be to use n.iter to activate rarefaction, and if it is activated then the rarefaction depth must be defined in any case and min(readcounts) is a standard default. 

This might have been requested to be NULL before we had the n.iter and multiple rounds added, at that time it made sense as it was the only rarefaction parameter.

Now reconsidering this, how about the following:

- n.iter=0 by default -> no rarefaction (same result would occur if n.iter=NULL or n.iter=FALSE)
- n.iter>0 (should be integer; should be checked) -> rarefaction + rarefaction.depth would be applied  

rarefaction.depth default value min(colSums(assay(tse, assay.type))) and this is applied only when n.iter>0

This procedure shuld be explained in the argument description and in @details field

This would be intuitive for a user?

ChouaibB replied:
Everything is fine by me. Initially the default for rarefaction.depth was that way.
Alright, I re-implement it.

antagomir replied:
Yes - but this does not seem to require full reimplementation, mainly changing the argument default values and how to deal with them.

TuomasBorman replied:
`n.iter` controls whether rarefaction is applied. `sample` is passed with `...` to `rarefyAssay()` so it is not handled in this function

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#' @param rarefaction.depth a \code{double} value as for the minimim size or 
+#'    rarefaction.depth. (By default: \code{rarefaction.depth = NULL})
+#' 
+#' @return \code{x} with additional \code{\link{colData}} named after the index 
```
I think `\code{x}` should be replaced by `\code{\link{SummarizedExperiment}}`?

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#'
+#' # Calculate observed richness with 10 rarefaction rounds
+#' tse <- addAlpha(tse,
+#'  assay.type = "counts",
+#'  index = "observed_richness",
+#'  rarefaction.depth=min(colSums(assay(tse, "counts")), na.rm = TRUE),
+#'  n.iter=10)
```
indentations should be multiples of 4 spaces (Bioc guidelines)

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#' @importFrom DelayedMatrixStats colMeans2
+.alpha_rarefaction <- function(
+        x, assay.type, n.iter, rarefaction.depth, FUN, index, name, ...){
+    # Calculating the mean of the subsampled alpha estimates ans storing them
```
ans -> and

antagomir commented on R/addAlpha.R:
```diff
@@ -0,0 +1,234 @@
+#'   
+#' @param ... optional arguments.
+#' 
+#' @param n.iter a single \code{integer} value for the number of rarefaction
+#'    rounds (By default: \code{n.iter = 10}).
```
By default rarefaction should not be applied.

If n.iter=1 then rarefaction should be applied once at the given rarefaction depth. That would be the standard single-round rarification.

By default n.iter should be switched off. This could be n.iter=0, or n.iter=FALSE, or n.iter=NULL. Perhaps the last one n.iter=NULL would be the least ambiguous. Also n.iter=0 should lead to the same result, calculating alpha diversity without any rarefaction.

Add here the following explanation:  "Rarefaction is not applied by default (n.iter=NULL); if n.iter is a positive integer then rarefaction is applied (see @details section)." -> Then in @details explain how n.iter and rarefaction.depth operate.

antagomir replied:
Unit tests should then check those conditions as well.

antagomir reviewed (commented):
Looks very good for now.

antagomir commented:
Some remaining comments + BiocChecks will be a must.

We should probably aim to merge this very soon, then do some further testing on the fresh new version.

antagomir commented:
Can we utilize the rarefyAssay function directly (see #536) - if not done yet?

TuomasBorman committed:
Sync with devel

Sync

# Conflicts:
#	DESCRIPTION
#	NAMESPACE
#	R/deprecate.R
#	R/estimateDiversity.R
#	R/getPrevalence.R
#	R/rarefyAssay.R
#	man/deprecate.Rd
#	man/estimateDiversity.Rd
#	man/estimateDominance.Rd
#	man/estimateEvenness.Rd
#	man/estimateRichness.Rd
#	man/mia-package.Rd
#	man/rarefyAssay.Rd
#	pkgdown/_pkgdown.yml
#	tests/testthat/test-0diversity.R
#	tests/testthat/test-8subsample.R

TuomasBorman committed:
Sync

Merge branch 'issue_417' of github.com:microbiome/mia into issue_417

# Conflicts:
#	R/deprecate.R
#	R/estimateDiversity.R
#	R/estimateDominance.R
#	R/estimateEvenness.R
#	man/deprecate.Rd
#	man/mergeSEs.Rd
#	pkgdown/_pkgdown.yml

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
Update estimateDiversity.R

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'issue_417' of github.com:microbiome/mia into issue_417

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
Bugfix

TuomasBorman committed:
Simplify

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman commented:
I think this is good to go. @ake123 Do you have time to check that there are necessary unit tests (there might be already)? Check Leo's comments about testing whether rarefying was disabled (n.iter = NULL and n.iter = 0)

TuomasBorman committed:
up

ake123 commented:
Sure I can check that

> I think this is good to go. @ake123 Do you have time to check that there are necessary unit tests (there might be already)? Check Leo's comments about testing whether rarefying was disabled (n.iter = NULL and n.iter = 0)

TuomasBorman committed:
Final fixes

TuomasBorman committed:
up

TuomasBorman committed:
update docs

TuomasBorman committed:
up

TuomasBorman commented on R/addAlpha.R:
```diff
@@ -0,0 +1,632 @@
+#'   specifying the links between rows and node labels of \code{tree}. If a
+#'   certain row is not linked with the tree, missing instance should be noted
+#'   as NA. When \code{NULL}, all the rownames should be found from the tree.
+#'   (Default: \code{NULL})
+#'   
+#'   \item \code{only.tips}: A boolean value specifying whether to remove
+#'   internal nodes when Faith's index is calculated. When
+#'   \code{only.tips=TRUE}, those rows that are not tips of tree are removed.
+#'   (Default: \code{FALSE})
+#'   
+#'   \item \code{threshold}: A numeric value in the unit interval,
+#'   determining the threshold for coverage and eveness indices.
+#'   (Default: \code{0.5})
+#'   
+#'   \item \code{quantile}: Arithmetic abundance classes are evenly cut up to to
+#'   this quantile of the data. The assumption is that abundances higher than
+#'   this are not common, and they are classified in their own group. This is
+#'   used for diversity indices.
+#'   (Default: \code{0.5})
+#'   
+#'   \item \code{nclasses}: The number of arithmetic abundance classes
+#'   from zero to the quantile cutoff indicated by \code{quantile}. This is used
+#'   for diversity indices. (Default: \code{50})
+#'   
+#'   \item \code{ntaxa}: Optional and only used for the \code{Absolute} and
+#'   \code{Relative} dominance indices: The n-th position of the dominant taxa
+#'   to consider. Disregarded for the indices \dQuote{dbp},
+#'   \dQuote{core_abundance}, \dQuote{Gini}, \dQuote{dmn}, and \dQuote{Simpson}.
+#'   (Default: \code{1})
+#'   
+#'   \item \code{aggregate}: Optional and only used for the \code{Absolute},
+#'   \code{dbp}, \code{Relative}, and \code{dmn} dominance indices.
+#'   Aggregate the values for top members selected by \code{ntaxa} or not. If
+#'   \code{TRUE}, then the sum of relative abundances is returned. Otherwise the
+#'   relative abundance is returned for the single taxa with the indicated rank
+#'   (default: \code{aggregate = TRUE}).
+#'   
+#'   \item \code{detection}: A \code{numeric} value for selecting detection
+#'   threshold for the abundances for richness indices. (Default: \code{0})
+#' }
```
TODO: What these parameters do? Which parameter affects which index?

TuomasBorman committed:
update docs

TuomasBorman commented:
Fixed some issues and updated documentation. Would be nice to have someone else's thoughts (if everything works) but I think this is good now

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman commented:
One last thing: I changed `n.iter` to `niter` to harmonize the paramater names. `nsomething` is used to denote number of something. We have already `nclasses` and `ndimred`.

If someone is against this change, I'm happy to discuss about this

antagomir reviewed (approved):
Looks good!

ake123 commented:
It seems the unit test have been updated with niter=NULL, niter=0 as default and with no errors.

TuomasBorman commented:
@ake123 Can you check that OMA and mia* packages are updated?

ake123 commented:
Yes sure! Should I add section addAlpha in the diversity section?
